feat/cli: support --log-dir arg to optionally log to a file instead of stdout/stderr

bochaco opened the pull request.

madadam commented on tests/common/mod.rs:
```diff
@@ -67,8 +66,6 @@ impl Environment {
     pub fn with_multiple_vaults(num_vaults: usize) -> Self {
         assert!(num_vaults > 0);
 
-        logging::init();
```
Does this mean we no longer log in tests? That could make debugging difficult.

maqi replied:
yeah, same doubt here

bochaco replied:
I'll investigate to make sure logs still work on tests

bochaco replied:
You were right, I just fixed that

madadam commented on src/bin/safe_vault.rs:
```diff
@@ -66,27 +66,41 @@ mod detail {
-                config.verbose().to_level_filter(),
-            );
+
+        // Let's filter out logs which are not from this crate, unless level chosen is ERROR
```
This might not be ideal. Sometimes it can be helpful to see the other crate logs when debugging. For example in testnet when something goes wrong the actual error can be anywhere in our stack and this would make it hard to track.

maqi replied:
agreed. maybe better make this as an configurable option as well.

bochaco replied:
I agree, I was simply trying to keep the same logic it had, from my understanding that's what previous `logger.filter(...)` was doing, and it can be overriden with RUST_LOG, but yeah, I can remove the filter to this crate's only

maqi commented on src/bin/safe_vault.rs:
```diff
@@ -66,27 +66,41 @@ mod detail {
-            writeln!(
-                formatter,
+        // Custom formatter for logs
+        let do_format = move |write: &mut dyn Write, now: &mut DeferredNow, record: &Record| {
```
these `write` `now` is a bit keyword like, maybe better use different variable name?
say `writer` and `clock` ?

bochaco replied:
yeah, true, I'll rename them

maqi commented on src/bin/safe_vault.rs:
```diff
@@ -66,27 +66,41 @@ mod detail {
-            );
+
+        // Let's filter out logs which are not from this crate, unless level chosen is ERROR
+        let verbosity = config.verbose().to_level_filter();
```
```rust
let filter = {
    let verbosity = config.verbose().to_level_filter();
    let mut filter = verbosity.to_string();
    if verbosity != Level::Error {
        if let Some(crate_name) = Config::clap().get_bin_name() {
            filter = format!("{}={}", crate_name, verbosity);
        }
    }
    filter 
};
```

maqi commented on tests/common/mod.rs:
```diff
@@ -67,8 +66,6 @@ impl Environment {
     pub fn with_multiple_vaults(num_vaults: usize) -> Self {
         assert!(num_vaults > 0);
 
-        logging::init();
```
just to confirm, this is removed because no longer required?

bochaco replied:
It's required, I just fixed that, thanks @maqi

S-Coyle commented:
I'm not sure why it's not being picked up by the Files tab diff, but in your `Cargo.toml` the `version` needs to be updated to v0.20.1 to match current upstream/fleming.
The result at the moment is that if a vault from this branch is launched it auto updates and the changes from this branch are lost in the updated running vault.

bochaco commented:
> I'm not sure why it's not being picked up by the Files tab diff, but in your `Cargo.toml` the `version` needs to be updated to v0.20.1 to match current upstream/fleming.
> The result at the moment is that if a vault from this branch is launched it auto updates and the changes from this branch are lost in the updated running vault.

I have no idea, but I've rebased it now with fleming branch, self-update won't happen anyway as I'm also disabled it and added a `--update` flag.

I think moving forward we should also migrate to use https://crates.io/crates/structopt , based on the experience in CLI is much simpler and a cleaner code.
